### PR TITLE
[NTR-141] Uploading csv to s3

### PR DESCRIPTION
**Pull request by likeuwill**
## Motivation and Context

[JIRA](https://eaflood.atlassian.net/browse/NTR-141)
## Description

As a Licensing AuthorityI want to upload a CSV fileSo that I may update the National Taxi Register
## How Has This Been Tested?

Added cucumber and rspec tests
## Checklist:

- [ ] It contains only changes required by issue (does not contain other PR)
- [ ] Includes link to a issue (if apply)
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by miqs1992** [comment]
Please rebase

**Comment by likeuwill** [comment]
> Please rebase
@miqs1992 
done

File: .rubocop.yml [hunk 1]
```diff
@@ -7,12 +7,18 @@ AllCops:
...
+Style/IfUnlessModifier:
+  Enabled: false
+
 Metrics/LineLength:
   Max: 99
   IgnoredPatterns: ['\A#']
 
 Metrics/MethodLength:
-  Max: 21
+  Max: 23
```
Comment (miqs1992) [inline .rubocop.yml]: refactor method or add local `rubocop:disable Metrics/MethodLength` rather than increase global attribute

Comment (miqs1992) [reply .rubocop.yml]: in `RemoteAuthenticatable.remote_authentication` you can move part with `client.initiate_auth` to separate method

File: app/services/csv_upload_base_service.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+class CsvUploadBaseService < BaseService
```
Comment (miqs1992) [inline app/services/csv_upload_base_service.rb]: I don't really see a point in dividing this service into 3 separate services - this one is just a dispatcher

File: app/services/base_service.rb [hunk 3]
```diff
@@ -1,7 +1,14 @@
 # frozen_string_literal: true
 
 class BaseService
+  attr_reader :file, :errors
```
Comment (miqs1992) [inline app/services/base_service.rb]: `BaseService` should be as generic as possible - not every service has a `:file` attribute. Initializer can easily be moved to actual service.

File: app/services/base_service.rb [hunk 4]
```diff
@@ -1,7 +1,14 @@
...
+
   def self.call(**args)
-    new(args).call
+    new(args).execute
```
Comment (miqs1992) [inline app/services/base_service.rb]: What was wrong in `call`? In `BaseService` it is a class method but in other ones, it is an instance method

File: app/views/layouts/_navbar.haml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+%button.govuk-header__menu-button.js-header-toggle{"aria-controls" => "navigation",
+  "aria-label" => "Show or hide Top Level Navigation", :role => "button", :type => "button"} Menu
```
Comment (miqs1992) [inline app/views/layouts/_navbar.haml]: use `role: 'button'` format

File: config/initializers/remote_authenticatable.rb [hunk 6]
```diff
@@ -25,20 +26,25 @@ def remote_authentication(params)
...
         rescue Aws::CognitoIdentityProvider::Errors::NotAuthorizedException
-          #TODO: Log errors
+          # TODO: Log errors
           false
         rescue StandardError
           false
         end
       end
 
+      def parse_user_email(response)
```
Comment (miqs1992) [inline config/initializers/remote_authenticatable.rb]: `parse_user_email` suggests returning an email while it returns a user.

File: app/services/validate_csv_service.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class ValidateCsvService < BaseService
+  FILENAME_FORMAT = /CAZ-([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))-([a-zA-Z]+)-\d+/.freeze
```
Comment (miqs1992) [inline app/services/validate_csv_service.rb]: It would be much more readable if you would first split name by '-' and than validate each piece

File: app/views/devise/sessions/new.html.haml [hunk 8]
```diff
@@ -1,9 +1,7 @@
 %main.govuk-main-wrapper#main-content{role: 'main'}
   %fieldset.govuk-fieldset
   %legend.govuk-fieldset__legend.govuk-fieldset__legend--xl
-    %h1.govuk-fieldset__heading
-      = t('.sign_in')
-    %br
+    %h1.govuk-heading-l Sign In
```
Comment (miqs1992) [inline app/views/devise/sessions/new.html.haml]: Please, do such a change in a separate commit. This one changes 42 files and it is extremely difficult to review. Changing the size of login input does not relate to file uploading.

File: config/initializers/remote_authenticatable.rb [hunk 9]
```diff
@@ -1,5 +1,6 @@
 # frozen_string_literal: true
 
+# :nocov:
```
Comment (miqs1992) [inline config/initializers/remote_authenticatable.rb]: Why this is excluded? Is it not touched during login tests?

File: app/services/csv_upload_base_service.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  def execute
+    if validate.errors || upload_to_s3.errors
+      @errors = validate.errors || upload_to_s3.errors
+    end
+    self
```
Comment (miqs1992) [inline app/services/csv_upload_base_service.rb]: Service should return an object if it is used for fetching something or just `true` if it is successful. It would be better to raise exceptions with an error message than return whole service object - this is the point of using `.call` class variable

**Comment by likeuwill** [comment]
Changes made. I have question only about `FILENAME_FORMAT`

**Comment by miqs1992** [comment]
We are very close - you can remove `Metrics/MethodLength: Max: 21` from .rubocop.yml if you divide `sign_in_user` in `features/step_definitions/helper_steps.rb` into:
```
def sign_in_user
  visit new_user_session_path
  fill_in('user_username', with: 'user@example.com')
  fill_in('user_password', with: '12345678')

  RSpec::Mocks.with_temporary_scope do
    expect_any_instance_of(Aws::CognitoIdentityProvider::Client).to receive(:initiate_auth)
      .and_return(cognito_response)

    click_button 'Continue'
  end
end

def cognito_response
  OpenStruct.new(challenge_parameters: {
                   'USER_ID_FOR_SRP' => 'user@example.com',
                   'userAttributes' => '{"email":""}'
                 })
end
```

**Comment by likeuwill** [comment]
@miqs1992 Done! ;)
